Fix: Allow irregular whitespace in comments. Fixes #5368.

tdd opened the pull request:
The `no-irregular-whitespace` rule is unduly limiting "irregular"
whitespace to string literals alone (despite a comment at the top
of the file indicating otherwise).  This prevents non-English
speakers from commenting their code in their own language.  For
instance, French language mandates (half-width, but hey) unbreakable
whitespace before two-part punctuation marks (in effect: `;`, `:`,
`?` and `!`), and this rule prevents them from doing so.

This commit fixes that.  Technically, it required choosing a different
parsing path as the existing use cases: apparently the rule
infrastructure only triggers `'Program`' entities after any potential
top-of-file comment block (or perhaps Esprima does this), and because
of the way the `no-irregular-whitespace` rule is originally written,
getting an initial `LineComment` or `BlockComment` node before the
main `Program` node prevents us from then stripping these errors
from the report.  So we accrue comment nodes until `Program:exit`,
then use them for error stripping at that final moment, right before
contributing to the report.

This also updates the docs and tests (all tests passing, checked).

BYK commented:
Hey there. You have two commits in your diff and none of them conforms to the required format. Can you run `git rebase -i master`, squash the second commit, make the commit message `Fix: Allow irregular whitespace in comments (fixes #5368)` and `git push -f`?

tdd commented:
@BYK yeah noticed the output from CI, just did that. Looking forward to these things being green!

tdd commented:
:tada: :+1: :grin:

BYK commented on lib/rules/no-irregular-whitespace.js:
```diff
@@ -18,13 +18,20 @@ module.exports = function(context) {
+
+    // Lookup the `skipComments` option, which defaults to `true`.
+    var options = context.options[0] || {};
+    var skipComments = typeof options.skipComments === "undefined" ? true : !!options.skipComments;
```
We didn't have this option before so I think making it "on" by default is a breaking change, thus a bad idea.

tdd replied:
I get this, but as the original intent was to preserve ES5 parsers (that don't care about irregular whitespace in comments), and as the comment at the beginning of the original code file stated "in comments too", I figured this made sense.

If that's truly a problem in your book, I can revert that. Then I'll ask StandardJS to have it on by default :wink:

BYK replied:
Deferring to @ilyavolodin and @gyandeeps for the final decision :)

ilyavolodin replied:
Hmm.. This is an interesting problem. I would say `on` by default is desired behavior, but it is breaking change (it will stop flagging errors that it flagged before). I'm not sure what's the right way here. @nzakas are you ok with this going in with on by default, but not marked as breaking?

nzakas replied:
I think we have to do off by default for back compat. I think wanting irregular whitespace in comments is the exception case, so starting with it off makes sense. If we had more feedback in the future, we can evaluate if it's with a breaking change at that point.

tdd commented:
Hey @nzakas and team!

OK, so I switched this around to defaulting to `false` (and updated docs and tests). All checks pass. Any idea when this could get released?

BYK commented on lib/rules/no-irregular-whitespace.js:
```diff
@@ -18,13 +19,20 @@ module.exports = function(context) {
+
+    // Lookup the `skipComments` option, which defaults to `false`.
+    var options = context.options[0] || {};
+    var skipComments = typeof options.skipComments === "undefined" ? false : !!options.skipComments;
```
You can now simply do `options.skipComments` here.

nzakas replied:
Rather `!!options.skipComments` will work.

tdd replied:
Dang.  Right, it can be simplified now.  Just pushed an updated PR.

--
Christophe Porteneuve
tdd@tddsworld.com

BYK commented:
Hey @tdd thanks for changes and keeping up with the requests! :)

Once this is merged, it will go out with the next release. [We do releases on Fridays or Saturdays every two weeks](https://github.com/eslint/eslint#releases). Last one was last week so the next release will be next week.

ilyavolodin commented:
LGTM

ilyavolodin commented:
@tdd One more question, did you sign our CLA? If not, could you do that? The link is in the bot's comments.

githubtrainer commented:
Thank you @BYK for the clarification.

@ilyavolodin yes I did sign it, first thing. Don't you team have a way to check that by looking at the Google Form results?

tdd commented:
*(dang, was logged in using a training account)*

Thank you @BYK for the clarification. So that's expecting this PR to be merged in-between…

@ilyavolodin yes I did sign it, first thing. Don't you team have a way to check that by looking at the Google Form results?

ilyavolodin commented:
@tdd Unfortunately due to security reasons, only @nzakas has access to that form. So we have to rely on him to check for CLA signatures (maybe we should change that, not sure). Anyways, thanks for PR!
